### PR TITLE
Database integrated in browser

### PR DESCRIPTION
**Pull request by annierusu**
Database is linked to the browser with the additional property for the card to be owned or wanted.

### CODE REVIEW

**Review by Cocograal — APPROVED** [review]
All good

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt [hunk 1]
```diff
@@ -51,7 +46,7 @@ class UserCardsViewModel : ViewModel() {
      * Add the card to the user's collection
      */
     fun onSetButtonClick(card: MagicCard) {
-        usc.addCardToCollection(card)
+        usc.addCardToCollection(FBMagicCard(card, FBCardPossession.NONE))
```
Comment (annierusu) [inline app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt:49]: possession should be owner for this add card

File: app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+    private val _buttonAddText = MutableLiveData<Boolean>()
+    val buttonAddText: LiveData<Boolean> = _buttonAddText
+
+    private val _buttonWantedText = MutableLiveData<Boolean>()
+    val buttonWantedText: LiveData<Boolean> = _buttonWantedText
+
+    private var userDB = UserCardsRTDB()
+
+    fun checkUserConnected() {
+        //TODO : Check if the user is connected
```
Comment (annierusu) [inline app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardViewModel.kt:31]: is this done?

File: app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardFragment.kt [hunk 3]
```diff
@@ -40,29 +79,68 @@ class SingleCardFragment: Fragment() {
...
         } catch (e: Exception) {
             binding.singleCardTextCardName.text = getString(R.string.error_load_card)
         }
     }
 
+    /**
+     * Displays the buttons to add the card to the deck or to the wanted cards if the user is connected.
+     * @param boolean true if the user is connected, false otherwise.
+     */
+    private fun displayButton(boolean: Boolean) {
```
[the remark this answers is not in the document]
Comment (KelvinKappeler) [reply app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardFragment.kt:132]: In addition, the comment implies that this boolean should be changed to true if the user is connected. That's fine, but this function doesn't know the user. It's just a function that shows or not some buttons. I don't remember how we called that in SDP, it was implementation leaking ?

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt [hunk 4]
```diff
@@ -60,15 +55,15 @@ class UserCardsViewModel : ViewModel() {
      * Remove the card from the user's collection
      */
     fun onRemoveButtonClick(card: MagicCard) {
-        usc.removeCardFromCollection(card)
+        usc.removeCardFromCollection(FBMagicCard(card, FBCardPossession.NONE))
```
Comment (KelvinKappeler) [inline app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt:58]: Why using a fake _FBCardPossession.NONE_ if you won't use it in _removeCardFromCollection_ ?

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt [hunk 5]
```diff
@@ -18,38 +15,47 @@ import java.util.concurrent.CompletableFuture
...
+        userCardCollection?.child(cardUID)?.setValue(data)
     }
 
     /**
      * Removes a card from the user's collection.
      */
-    fun removeCardFromCollection(card: MagicCard) {
-        val cardUID = card.set.code + card.number
-        userCardCollection.child(cardUID).removeValue()
+    fun removeCardFromCollection(fbCard: FBMagicCard) {
```
Comment (KelvinKappeler) [inline app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt:40]: Why using FBMagicCard here? We don't care about the possession of a card.

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt [hunk 6]
```diff
@@ -18,38 +15,47 @@ import java.util.concurrent.CompletableFuture
...
+        userCardCollection?.child(cardUID)?.removeValue()
     }
 
     /**
      * Retrieves a card asynchronously from the database
+     * The card is identified by only its set code and its number
      */
-    fun getCardFromCollection(card: MagicCard): CompletableFuture<DataSnapshot> {
-        val cardUID = card.set.code + card.number
+    fun getCardFromCollection(fbCard: FBMagicCard): CompletableFuture<DataSnapshot> {
```
Comment (KelvinKappeler) [inline app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt:49]: Here too.

**Review by Cocograal — DISMISSED** [review]

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt [hunk 7]
```diff
@@ -60,15 +55,15 @@ class UserCardsViewModel : ViewModel() {
...
         _removeMessage.value = Pair(R.string.text_remove_success, card.name)
     }
 
     /**
      * Get the card from the user's collection if it exists
      */
     fun onGetButtonClick(card: MagicCard) {
-        val data = usc.getCardFromCollection(card)
+        val data = usc.getCardFromCollection(FBMagicCard(card, FBCardPossession.NONE))
```
Comment (KelvinKappeler) [inline app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt:66]: Again same comment.

File: app/src/main/res/layout/fragment_single_card.xml [hunk 8]
```diff
@@ -125,5 +125,36 @@
...
+            android:layout_marginTop="10sp"
+            android:layout_marginBottom="10sp">
+
+            <Button
+                android:id="@+id/singleCard_button_addToDeck"
+                android:layout_width="wrap_content"
+                android:layout_height="match_parent"
+                android:layout_gravity="center_horizontal"
+                android:text="@string/single_card_showing_add_to_collection"
+                android:layout_marginRight="10dp" />
```
Comment (KelvinKappeler) [inline app/src/main/res/layout/fragment_single_card.xml:141]: You should use android:layout_marginEnd="10dp", not android:layout_marginRight

**Review by KelvinKappeler — CHANGES_REQUESTED** [review]
Good work in general, it should be discussed with the team, but I find that there are a lot of unnecessary parameters (having to add a possession by hand when it is useless for example).

In addition, the tests are sorely lacking (41% is not enough).

Il will wait for another review from our team :)

File: app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardFragment.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (annierusu) [reply app/src/main/java/com/github/sdp/tarjetakuna/ui/singlecard/SingleCardFragment.kt:132]: The argument in this function is an observer (line 40) of isConnected inside the viewmodel, which does have access to the connectivity of the user. Can you elaborate what you mean?

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt (hunk 4 above)
Comment (annierusu) [reply app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsViewModel.kt:58]: yes, I just realised kotlin can have default arguments, which we should use, we keep every argument type as an FB card to keep the database code consistent.

**Review by jeamigue — DISMISSED** [review]
Some improvements could be made as pointed out by the other reviewer. I suggest we merge and create a new branch to improve the adressed issues, so we don't have big changes pending for too long, since nothing really blocking. Perfect should not always be the target.

File: app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt (hunk 5 above)
Comment (annierusu) [reply app/src/main/java/com/github/sdp/tarjetakuna/database/UserCardsRTDB.kt:40]: We use FBMagicCard in all the methods of the database, this is to keep the code consistent.

**Review by KelvinKappeler — APPROVED** [review]
Thanks for your changes.

**Pull request merged**